### ISSUE TITLE
Site search research

### ISSUE THREAD

**Issue by desigonz** [cityofaustin/techstack#4113]
What does it look like to search a statically generated site? What's out there?

**Comment by robbobfrh84** [comment]
I starting looking into this and found some interesting routes. Here's my initial unorganized 🧠dump 🚮...
 
Elastic Search with React
- https://www.elastic.co/blog/how-to-build-application-search-with-react-and-elastic-app-search
- Here's Tutorial Guide to creating an example app: https://codeburst.io/how-to-build-great-react-search-experiences-quickly-8c69081f328d
- Question about this route... Does react-static change expected behavior?

Elastic search isn't the only route - just the first that came to mind - there may be other options. This looked interesting and maybe a little simpler to fit our needs...
- https://survivejs.com/webpack/appendices/searching-with-react/

### OK, but... 

What does a resident want / expect from a "search" functionality?
- As I started digging into this, I realized that this could be a simple "pages" search (like what we do on joplin?), or something more complex that doesn't just search for pages, but filters and handles ranges, etc... etc.. 

Would it be reasonable to bring some content perspective to this convo?
- I can see how a little "what" we want and "why" we want it might make the direction we go on development more clear.

**Comment by easherma** [comment]
At the heart of any search solution is going to be the ability to have a site map, build a searchable index of the website, and have an API to be able to query against.

There are many solutions out there that assist with this. Algolia is one that comes to mind, Solr is another. They are going to require setting up a site index, which we haven’t really done yet, and will be interesting to try to set up given how reluctant we’ve been to adopt a tree-like structure.

Elasticsearch is a great option that takes a good amount of setup(I’ve set up Elasticsearch severs for clients in the past)

There is also search functionality that Wagtail provides that we could potentially leverage against their REST API.

I think Bob is right that we’d want specific guidance on what kind of searching we want to support. A good MVP approach would probably be to leverage an existing service so that time and resources could be spent exploring this question and building the site index and not on building infrastructure  that may or may not suit the project.

**Comment by Niicck** [comment]
TL;DR: I'd vote for using Algolia as a good enough MVP for us.
- it works out of the box without a ton of configuration or effort. This will at least get us rolling. 
- There's no reason why we couldn't switch over to Elasticsearch later on if we decide we need greater control over our search queries.

Elasticsearch is used by gov.uk.
- Old article when elasticsearch was first implemented: https://insidegovuk.blog.gov.uk/2014/06/13/how-gov-uk-site-search-works/
- A recent article about how they're improving it: https://insidegovuk.blog.gov.uk/2019/07/05/making-site-search-work-smarter-for-users/
- An important thing to note is the gov.uk has an entire team dedicated to designing and developing better search with elasticsearch.

Algolia is basically a plug and play search service. If just need some kind of acceptable site search, Algolia will do the job. If search becomes a higher priority for us and we want to seriously devote dev, design, and management resources to it, we'd probably switch over to Elasticsearch. But even after getting Elasticsearch set up, its not a light switch that magically makes search better. There would still be a lot of effort to accurate tune our search algorithm.

So what are our requirements? How intelligent do we want to be right now?
- Do we want....
  - filtering by department
  - filtering by content type
  - adding human-readable "tags" and "synonyms" to pages
- And then how pretty do we want the user experience to be for content authors tweaking search terms and queries?
  - Elasticsearch will give us more tools than Algolia will. (Bob's link shows us what that could look like: https://codeburst.io/how-to-build-great-react-search-experiences-quickly-8c69081f328d)
  - Algolia's more advanced tools are locked behind a paywall.
  - Algolia is also limited to a set number of users.
  - Which users should be allowed to tweak the search algorithm? Should any?

How Elasticsearch/Algolia work:
  - You send them data (e.g. a page)
  - They store that page and add the words inside of it to an inverted-index. That means when you search for a term (like "recycling"), the search engine will instantly be able to return all the pages that contain the word "recycling".
  - You send a query from Janis's search page to the search service, and that service returns a list of pages that contain your search term.

How Elasticsearch would probably work for us:
  - We'd have to set up an elasticsearch database. They are hosted/managed solutions out there, but apparently its still a hassle to set up and maintain. This is one of the major selling points of Algolia.
  - A library like this (https://github.com/sabricot/django-elasticsearch-dsl) would automagically sync pages into elasticsearch from Joplin as they're updated.
  - Content authors would use the interface that Bob linked to: https://codeburst.io/how-to-build-great-react-search-experiences-quickly-8c69081f328d. That there "Elastic App Search" looks like the right product we'd want to use to manage search configurations.

How Algolia would probably work for us:
- We set up Algolia. This apparently takes like 45 minutes and is super easy.
- They too have a nice django library. It's even officially supported: https://www.algolia.com/doc/framework-integration/django/setup/?language=python


Cost of Algolia
- https://www.algolia.com/pricing/
- It's free for 50K operations and 10K records. In our use case, records would be pages. This would work for PR and staging.
- For production (at some point), we'd have to use a $29 "Starter" paid plan that includes 250K operations and 50k records. The reason for this is that the free plan would cap us at 50k operations. The Starter plan would allow us to buy extra search operations for $10 for every 100k operations.
- The cost difference between algolia and elasticsearch isn't as steep as it is for use cases involving millions of documents.

Additional references:
- A fun technical video about how the elasticsearch architecture works: https://www.youtube.com/watch?v=C3tlMqaNSaI
- Arguing the algolia is the right choice for small teams (that's us!) who only need basic string search functionality (still us?) https://opensourceconnections.com/blog/2016/06/01/thoughts-on-algolia/
- An algolia-biased, though informative, thread about the right use cases for algolia v. elasticsearch: https://www.quora.com/How-does-Elasticsearch-relate-and-or-compare-to-Algolias-Search-as-a-service
- A salty nerd debate about algolia v. elasticsearch: https://news.ycombinator.com/item?id=21257529

**Comment by desigonz** [comment]
dang @Niicck this is good!! looking forward to discussing tomorrow.

**Comment by briaguya** [comment]
@Niicck do we want Algolia to work with Joplin? If we're just searching content that's in Janis should our index be of the Janis build instead?

**Comment by Niicck** [comment]
That's open for discussion. I feel like it makes more sense for us to work with Joplin data directly, rather than the Janis build. The Joplin data directly has more intelligence behind it (topics, departments, updated datetime). If we index the Janis site itself, there's more automated guesswork about what's important and relevant.

**Comment by chiaberry** [comment]
Is there content that exists in Janis but does not exist in Joplin? I am thinking its only context that is included in Janis, such as headings and other labels we translate, but wanted to bring it up in case there was something else that exists in Janis and not in the data.

**Comment by desigonz** [comment]
@chiaberry Good question! I don't think there's any content we'd want to index for search that isn't in the database, but I could be wrong.
